data/src/test/java/com/linkedin/data/codec/entitystream/TestProtobufDataDecoder.java

```diff
@@ -0,0 +1,184 @@
+      DataMap decodedMap = (DataMap) decode(bytes);
+      assertEquals(decodedMap.getInteger("int"), Integer.valueOf(i));
+    }
+  }
+
+  @Test
+  public void testLongValues() throws Exception
+  {
+    long longInc = (Long.MAX_VALUE - Long.MAX_VALUE / 100L) / 10000L;
+    for (long i = Long.MAX_VALUE / 100L; i <= Long.MAX_VALUE && i > 0; i += longInc)
+    {
+      DataMap dataMap = new DataMap();
+      dataMap.put("long", i);
+      byte[] bytes =
+          TestUtil.dataComplexToBytes(new ProtobufDataCodec(new ProtobufCodecOptions.Builder().build()), dataMap);
+      DataMap decodedMap = (DataMap) decode(bytes);
+      assertEquals(decodedMap.getLong("long"), Long.valueOf(i));
+    }
+    for (long i = Long.MIN_VALUE; i <= Long.MIN_VALUE / 100L && i < 0; i += longInc)
+    {
+      DataMap dataMap = new DataMap();
+      dataMap.put("long", i);
+      byte[] bytes =
+          TestUtil.dataComplexToBytes(new ProtobufDataCodec(new ProtobufCodecOptions.Builder().build()), dataMap);
+      DataMap decodedMap = (DataMap) decode(bytes);
+      assertEquals(decodedMap.getLong("long"), Long.valueOf(i));
+    }
+  }
+
+  @Test
+  public void testInvalidMap() throws Exception
+  {
+    DataList dataList = new DataList();
+    dataList.add(1);
+    dataList.add(2);
+    dataList.add(4);
+    byte[] bytes =
+        TestUtil.dataComplexToBytes(new ProtobufDataCodec(new ProtobufCodecOptions.Builder().build()), dataList);
+    decode(bytes);
+
+    try
+    {
+      decodeMap(bytes);
+      fail("Parsing list as map.");
+    }
+    catch (ExecutionException e)
+    {
+      // Expected.
+    }
+  }
+
+  @Test
+  public void testInvalidList() throws Exception
+  {
+    DataMap dataMap = new DataMap();
+    dataMap.put("key", true);
+    byte[] bytes =
+        TestUtil.dataComplexToBytes(new ProtobufDataCodec(new ProtobufCodecOptions.Builder().build()), dataMap);
+    decode(bytes);
+
+    try
+    {
+      decodeList(bytes);
+      fail("Parsing map as list");
+    }
+    catch (ExecutionException e)
+    {
+      // Expected.
+    }
+  }
+
+  private static DataComplex decode(byte[] bytes) throws Exception
+  {
+    ProtobufDataDecoder<DataComplex> decoder = new ProtobufDataDecoder<>(null, AbstractDataDecoder.START_TOKENS);
+    return decode(bytes, decoder);
+  }
+
+  private static DataMap decodeMap(byte[] bytes) throws Exception
+  {
+    ProtobufDataDecoder<DataMap> decoder =
+        new ProtobufDataDecoder<>(null, AbstractDataDecoder.START_OBJECT_TOKEN);
+    return decode(bytes, decoder);
+  }
+
+  private static DataList decodeList(byte[] bytes) throws Exception
+  {
+    ProtobufDataDecoder<DataList> decoder =
+        new ProtobufDataDecoder<>(null, AbstractDataDecoder.START_ARRAY_TOKEN);
+    return decode(bytes, decoder);
+  }
+
+  private static <T extends DataComplex> T decode(byte[] bytes, ProtobufDataDecoder<T> decoder) throws Exception
+  {
+    Writer<ByteString> writer = new ChunkedByteStringWriter(bytes, 3);
```

Comment:
Does the test pass if you set this to 1? That will be the ultimate litmus test since you would have accounted for all edge cases.

Reply:
will add test for different buffer size